sky/shell/mac/tracing_controller.mm

```diff
@@ -2,119 +2,130 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-#import "tracing_controller.h"
-#include <string>
+#include "base/logging.h"
 #include "base/macros.h"
 #include "base/trace_event/trace_config.h"
 #include "base/trace_event/trace_event.h"
-
-@interface TracingController ()
-@property(nonatomic, retain) NSFileHandle* currentFileHandle;
-@end
+#include "sky/shell/mac/tracing_controller.h"
+#include "sky/shell/shell.h"
+#include <Foundation/Foundation.h>
+#include <string>
 
 namespace sky {
 namespace shell {
 
-const char kStart[] = "{\"traceEvents\":[";
-const char kEnd[] = "]}";
+static const std::string kBaseTraceStart("{\"traceEvents\":[");
+static const std::string kBaseTraceEnd("]}");
+static const std::string kSentinel("\0", 1);
 
-static void Write(const std::string& data) {
-  NSFileHandle* handle = [TracingController sharedController].currentFileHandle;
-  [handle writeData:[NSData dataWithBytesNoCopy:(void*)data.data()
-                                         length:data.size()
-                                   freeWhenDone:false]];
+TracingController& TracingController::Shared() {
+  static std::unique_ptr<TracingController> instance;
```

Comment:
We can't have statics like this either because this creates an exit-time destructor.